metabase/metabase
# Add snowplow CI

Epic: https://github.com/metabase/metabase/issues/18291

- Moved snowplow schemas to this repository
- Added a job to lint the schemas
- Run `snowplow-micro` within the cypress executor for e2e testing
- Added a snowplow cypress test

**How to test**
- Check that `run-snowplow-micro` step is run for e2e tests
- Check that modified cypress tests pass
- Check that the lint job is executed without errors for existing schemas
- Check that the lint job fails if there are some issues with the schemas (e.g. remove `type` from any property)
- Check that `e2e-tests-onboarding` test suite fails on `expectGoodEvents(1)`

**Encountered problems**
- It's not possible to run `snowplow-micro` as a secondary executor, like `maildev`. The container requires volume mapping for access to the schemas, which is not supported by CircleCI.
- It's not possible to run `snowplow-micro` with `docker run` inside the primary executor as CircleCI doesn't not expose ports for inner containers. It's also not possible to remap `docker.sock` to avoid this issue.
- It's not possible to add `snowplow-micro` as a maven dependency since it's not published.
- It turned out that the only way is to download and run `snowplow-micro` as a JAR. The JAR is cached by CircleCI so it shouldn't be an issue.
